File: Java/src/main/java/parrot/Parrot.java
package parrot;

public abstract class Parrot {

    protected final boolean isNailed;

    protected Parrot(boolean isNailed) {
        this.isNailed = isNailed;
    }

    public abstract double getSpeed();

    protected double getBaseSpeed(double voltage) {
        return Math.min(24.0, voltage * getBaseSpeed());
    }

    protected double getLoadFactor() {
        return 9.0;
    }

    protected double getBaseSpeed() {
        return 12.0;
    }

    public abstract String getCry();
}

File: Kotlin/src/main/kotlin/parrot/Parrot.kt
package parrot

import kotlin.math.min

abstract class Parrot {

    protected val type: ParrotTypeEnum
    protected val numberOfCoconuts: Int
    protected val voltage: Double
    protected val isNailed: Boolean

    constructor(type: ParrotTypeEnum, numberOfCoconuts: Int, voltage: Double, isNailed: Boolean) {
        this.type = type
        this.numberOfCoconuts = numberOfCoconuts
        this.voltage = voltage
        this.isNailed = isNailed
    }

    protected val loadFactor: Double
        get() = 9.0

    protected val baseSpeed: Double
        get() = 12.0

    protected fun getBaseSpeed(voltage: Double): Double = min(24.0, voltage * baseSpeed)

    abstract val cry: String
    abstract val speed: Double
}
